onnxruntime/core/optimizer/common_subexpression_elimination.cc

```diff
@@ -0,0 +1,428 @@
+    case onnx::AttributeProto_AttributeType_TENSOR:
+    case onnx::AttributeProto_AttributeType_GRAPH:
+    case onnx::AttributeProto_AttributeType_SPARSE_TENSOR:
+    case onnx::AttributeProto_AttributeType_TENSORS:
+    case onnx::AttributeProto_AttributeType_GRAPHS:
+    case onnx::AttributeProto_AttributeType_SPARSE_TENSORS:
+    case onnx::AttributeProto_AttributeType_UNDEFINED:
+      break;
+  }
+
+  return hash;
+}
+
+bool SameAttribute(const NodeAttributes::value_type& lhs, const NodeAttributes::value_type& rhs) {
+  return lhs.first == rhs.first && AreEqual(lhs.second, rhs.second);
+}
+
+bool SameAttributes(const NodeAttributes* lhs, const NodeAttributes* rhs) {
+  if (lhs == nullptr || rhs == nullptr) {
+    return lhs == rhs;
+  }
+  return std::equal(lhs->begin(), lhs->end(), rhs->begin(), rhs->end(), &SameAttribute);
+}
+
+bool EquivalenceClass::operator==(const EquivalenceClass& other) const {
+  if (this == &other) {
+    return true;
+  }
+
+  // Below we compare inputs_ as pointers. This is valid due to how EquivalenceClass'es are constructed:
+  // we'll never have two distinct but equal inputs_ here, so their addresses are effectively their value numbers.
+  return hash_ == other.hash_ && output_index_ == other.output_index_ && discriminator_ == other.discriminator_ &&
+         non_op_value_ == other.non_op_value_ &&
+         op_type_ == other.op_type_ && domain_ == other.domain_ &&
+         inputs_ == other.inputs_ &&
+         SameAttributes(attributes_, other.attributes_);
+}
+
+bool EquivalenceClass::operator!=(const EquivalenceClass& other) const {
+  return !operator==(other);
+}
+
+std::size_t EquivalenceClass::CalculateHash() const {
+  std::size_t hash = 0;
+  UpdateHash(output_index_, hash);
+  UpdateHash(discriminator_, hash);
+  UpdateHash(non_op_value_, hash);
+  UpdateHash(op_type_, hash);
+  UpdateHash(domain_, hash);
+  if (attributes_ != nullptr) {
+    for (const auto& kv : *attributes_) {
+      UpdateHash(kv.first, hash);
+      UpdateHash(kv.second, &GetAttributeHash, hash);
+    }
+  }
+
+  for (const auto& arg : inputs_) {
+    for (const EquivalenceClass* input : arg) {
+      UpdateHash(input, DeepPointerHash{}, hash);
+    }
+  }
+
+  return hash;
+}
+
+// Representative of an equivalence class.
+// node_index and output_index define the node that produced the node_arg.
+// For inputs and constant initializers, output_index == kInvalidOutputIndex.
+struct Representative {
+  const NodeArg* node_arg;
+  NodeIndex node_index;
+  OutputIndex output_index;
+};
+
+struct NodeArgPtrHash {
+  std::size_t operator()(const NodeArg* node_arg) const {
+    return std::hash<const NodeArg*>{}(Normalize(node_arg));
+  }
+};
+
+struct NodeArgPtrEquality {
+  bool operator()(const NodeArg* lhs, const NodeArg* rhs) const {
+    return Normalize(lhs) == Normalize(rhs);
+  }
+};
+
+bool IsNodeSupported(const Node& node) {
+  if (node.Domain() != kOnnxDomain || node.ContainsSubgraph()) {
+    return false;
```

Comment:
What's the reason for excluding other domains?

Reply:
We don't know whether those operations are deterministic or not (e.g. custom ops)

Reply:
A comment saying that will help future developers looking at this and wondering why. 

Can we support the MS Domain given those ops are part of this codebase if contrib ops are enabled?

---
In reply to: [437070556](https://github.com/microsoft/onnxruntime/pull/3047#discussion_r437070556) [](ancestors = 437070556)

Reply:
As far as I know they are. I took a look and didn't notice any that weren't. Maybe just capture in a comment that these domains could potentially be enabled but leave as disabled for now. That will provide a hint if anyone is looking at this code and wondering why it doesn't work for those ops and they can carefully consider adding some/all at that point.